File: apps/api-python/src/neetcode_api/timing.py
"""A stopwatch scoped to the algorithm call only."""

from __future__ import annotations

import time
from collections.abc import Callable
from typing import Any, ParamSpec, TypeVar

P = ParamSpec("P")
R = TypeVar("R")


def timed(fn: Callable[P, R], *args: Any, **kwargs: Any) -> tuple[R, int]:
    """Run `fn` and return `(result, elapsed_microseconds)`.

    `perf_counter_ns` rather than `time()`: it is monotonic and has nanosecond resolution,
    which matters because the fast path of an O(n) solution on a six-element list finishes in
    well under a microsecond.
    """
    start = time.perf_counter_ns()
    result = fn(*args, **kwargs)
    return result, (time.perf_counter_ns() - start) // 1_000

File: apps/api-node/src/common/timing.ts
/**
 * A stopwatch scoped to the algorithm call only.
 *
 * `process.hrtime.bigint()` rather than `Date.now()`: it is monotonic and nanosecond-
 * resolution, which matters because an O(n) solution over six elements finishes in well
 * under a millisecond — `Date.now()` would report a flat 0 for every request.
 */
export function timed<T>(fn: () => T): { result: T; elapsedMicros: number } {
  const start = process.hrtime.bigint();
  const result = fn();
  return { result, elapsedMicros: Number((process.hrtime.bigint() - start) / 1000n) };
}
